### ISSUE TITLE
reduce vsl client stat() overhead

### ISSUE THREAD

**Issue by nigoroll** [varnishcache/varnish-cache#2788]
our `fstat()/fstatat()` frequency in the vcl client is way too high. Checking a `varnishncsa -q ... -F ...` process:

* from `strace -ttt`
```
1538140709.408295 nanosleep({0, 10000000}, NULL) = 0
1538140709.418572 fstat(3, {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
1538140709.418748 newfstatat(3, "_.vsm_mgt", {st_mode=S_IFDIR|0750, st_size=4096, ...}, AT_SYMLINK_NOFOLLOW) = 0
1538140709.418818 newfstatat(4, "_.index", {st_mode=S_IFREG|0640, st_size=291, ...}, AT_SYMLINK_NOFOLLOW) = 0
1538140709.418889 newfstatat(3, "_.vsm_child", {st_mode=S_IFDIR|0750, st_size=4096, ...}, AT_SYMLINK_NOFOLLOW) = 0
1538140709.418964 newfstatat(6, "_.index", {st_mode=S_IFREG|0640, st_size=9542, ...}, AT_SYMLINK_NOFOLLOW) = 0
1538140709.419041 fstat(3, {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
1538140709.419096 newfstatat(3, "_.vsm_mgt", {st_mode=S_IFDIR|0750, st_size=4096, ...}, AT_SYMLINK_NOFOLLOW) = 0
1538140709.419157 newfstatat(4, "_.index", {st_mode=S_IFREG|0640, st_size=291, ...}, AT_SYMLINK_NOFOLLOW) = 0
1538140709.419215 newfstatat(3, "_.vsm_child", {st_mode=S_IFDIR|0750, st_size=4096, ...}, AT_SYMLINK_NOFOLLOW) = 0
1538140709.419334 newfstatat(6, "_.index", {st_mode=S_IFREG|0640, st_size=9542, ...}, AT_SYMLINK_NOFOLLOW) = 0
1538140709.419424 fstat(3, {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
1538140709.419483 newfstatat(3, "_.vsm_mgt", {st_mode=S_IFDIR|0750, st_size=4096, ...}, AT_SYMLINK_NOFOLLOW) = 0
1538140709.419551 newfstatat(4, "_.index", {st_mode=S_IFREG|0640, st_size=291, ...}, AT_SYMLINK_NOFOLLOW) = 0
1538140709.419621 newfstatat(3, "_.vsm_child", {st_mode=S_IFDIR|0750, st_size=4096, ...}, AT_SYMLINK_NOFOLLOW) = 0
1538140709.419669 newfstatat(6, "_.index", {st_mode=S_IFREG|0640, st_size=9542, ...}, AT_SYMLINK_NOFOLLOW) = 0
```

* syscall rate
```
time strace -p ... 2>&1| head -1000 > /dev/null

real	0m0.712s
user	0m0.056s
sys	0m0.052s
```

**Comment by bsdphk** [comment]
This is not very much concrete information to go on...

Which program are you running ?

Is it using VUT ?

**Comment by nigoroll** [comment]
I had mentioned in the initial issue note that this is from `varnishncsa -q ... -F ...`

**Comment by bsdphk** [comment]
sorry, overlooked that.

**Comment by nigoroll** [comment]
I have just deployed a production environment, moving from e9e9355068f925afd965d4bab9aba9fcdefef77b to 
827cd226e42cd1048262b4260ee151eb602486ce (each plus two additional patches, see https://github.com/nigoroll/varnish-cache/tree/unmerged_code ).

With this deployment, we increased the number of VSL clients running from 7 to 8 (now 1x varnishlog, 7x varnishevent) yet still the load dropped by a good 20%. While I cannot exclude other possible causes, I am pretty sure that the improvement is related to this change.

this graph shows stacked 5-min load averages for 6 varnish servers. The spike is some erratic behavior during restart which is not yet understood

![load](https://user-images.githubusercontent.com/1528104/46727622-4bd68e80-cc81-11e8-9291-32e33d4d5005.png)
